### PR TITLE
CDAP-2025 Updated RunRecord of the Workflow and underlying MapReduce pro...

### PR DESCRIPTION
**Pull request by sagarkapare**
...grams.

Opening this PR against release branch.
Jira: https://issues.cask.co/browse/CDAP-2025

Summary of changes:
1. Passing the Workflow related information (Workflow name, Workflow RunId and node id of the program in the Workflow) to the MapReduceProgramRunner which uses this information to record and update the state for the MapReduce program and Workflow in the MDS.

2. Updated the RunRecord to include generic properties field which is of type Map. The value contained in the properties depends on the program for which the RunRecord is created.
e.g. For Workflow properties represents the map of node id to run id of the programs inside the Workflow. For MapReduce program which started as a part of the Workflow, properties currently represents map of single entry for the Workflow run id.

e.g. Assume that the `PurchaseHistoryWorkflow` has runs two MapReduce program node, each runs the same MapReduce program `PurchaseHistoryBuilder`.
Then the RunRecord for the Workflow would look -
```json
[
    {
        "end": 1429831538,
        "properties": {
            "0": "f3922dd4-ea0f-11e4-a768-c200d05e85ef",
            "1": "009b46b5-ea10-11e4-8e50-c200d05e85ef"
        },
        "runid": "f3807a93-ea0f-11e4-b6eb-c200d05e85ef",
        "start": 1429831495,
        "status": "COMPLETED"
    }
] 
```
For single run of the Workflow, there will be two RunRecords for the MapReduce program as -
```json
[
    {
        "end": 1429831538,
        "properties": {
            "workflowrunid": "f3807a93-ea0f-11e4-b6eb-c200d05e85ef"
        },
        "runid": "009b46b5-ea10-11e4-8e50-c200d05e85ef",
        "start": 1429831517,
        "status": "COMPLETED"
    },
    {
        "end": 1429831517,
        "properties": {
            "workflowrunid": "f3807a93-ea0f-11e4-b6eb-c200d05e85ef"
        },
        "runid": "f3922dd4-ea0f-11e4-a768-c200d05e85ef",
        "start": 1429831495,
        "status": "COMPLETED"
    }
] 
```

### CODE REVIEW

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-RBT222-1

**Comment by sagarkapare** [comment]
@anew @chtyim Please take a look. Thanks!

**Comment by sagarkapare** [comment]
Build passes here - http://builds.cask.co/browse/CDAP-RBT222-2

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java [hunk 1]
```diff
@@ -175,7 +178,13 @@ public void starting() {
           startTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
         }
         String adapterName = adapterSpec == null ? null : adapterSpec.getName();
-        store.setStart(program.getId(), runId.getId(), startTimeInSeconds, adapterName, twillRunId);
+        if (workflowName == null) {
+          store.setStart(program.getId(), runId.getId(), startTimeInSeconds, adapterName, twillRunId);
+        } else {
+          // Program started by Workflow
+          store.setWorkflowProgramStart(program.getId(), runId.getId(), workflowName, workflowRunId, workflowNodeId,
+                                        startTimeInSeconds, twillRunId);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: adapterName is not needed here?

Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: `adapterName` is fetched from the `Workflow` run record. However, for consistency its better to pass it from here?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/ProgramWorkflowRunnerFactory.java [hunk 2]
```diff
@@ -60,17 +60,18 @@ public ProgramWorkflowRunnerFactory(WorkflowSpecification workflowSpec, ProgramR
    * @param token      The {@link WorkflowToken}
    * @return the appropriate concrete implementation of {@link ProgramWorkflowRunner} for the program
    */
-  public ProgramWorkflowRunner getProgramWorkflowRunner(WorkflowActionSpecification actionSpec, WorkflowToken token) {
+  public ProgramWorkflowRunner getProgramWorkflowRunner(WorkflowActionSpecification actionSpec, WorkflowToken token,
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/ProgramWorkflowRunnerFactory.java]: doesn't a workflow action spec contain the node id? If not, why not?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 3]
```diff
@@ -548,4 +551,53 @@ private MDSKey getNamespaceKey(@Nullable String name) {
...
+    if (record == null) {
+      String msg = String.format("No meta found for associated Workflow %s run record %s, while recording run for the" +
+                                   " namespace %s app %s type %s program %s runid %s", workflow, workflowRunId,
+                                 program.getNamespaceId(), program.getApplicationId(), program.getType().name(),
+                                 program.getId(), programRunId);
+      LOG.error(msg);
+      throw new IllegalArgumentException(msg);
+    }
+
+    deleteAll(key);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: why delete all? please explain

File: cdap-proto/src/main/java/co/cask/cdap/proto/RunRecord.java [hunk 4]
```diff
@@ -44,14 +46,25 @@
...
   private final String adapterName;
 
+  @Nullable
+  @SerializedName("properties")
+  private final Map<String, String> properties;
+
   public RunRecord(String pid, long startTs, @Nullable Long stopTs, ProgramRunStatus status,
-                   @Nullable String adapterName, @Nullable String twillRunId) {
+                   @Nullable String adapterName, @Nullable String twillRunId,
+                   @Nullable Map<String, String> properties) {
```
Comment (anew) [inline cdap-proto/src/main/java/co/cask/cdap/proto/RunRecord.java]: It seems better to require that properties is not null, but rather an empty map if there is nothing there. That would be consistent with what we do for other similar parameters, such as RuntimeArguments or DatasetProperties.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 1 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Oh, ok. That make sense.

**Comment by anew** [comment]
@sagarkapare the approach is good, I have only a few comments.

**Comment by sagarkapare** [comment]
Thank you @anew for the review comments. Updating the PR as per suggestions.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 1 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Passing adapterName now.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/ProgramWorkflowRunnerFactory.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/ProgramWorkflowRunnerFactory.java]: `WorkflowActionSpecification` does not contain the node id. 

The hierarchy is, at the top level we have `WorkflowNode`, which contains `nodeId` and the `nodeType`. Action nodes are represented by the `WorkflowActionNode` class which inherits the `WorkflowNode` and include the `WorkflowActionSpecification` as well. 

Please let me know if nodeId should also be included in the action spec.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: Even if the method is named as `deleteAll`, the key formed here will cause the deletion of the single `RunRecord`entry corresponding to the runId specified by the parameters.

File: cdap-proto/src/main/java/co/cask/cdap/proto/RunRecord.java (hunk 4 above)
Comment (sagarkapare) [inline cdap-proto/src/main/java/co/cask/cdap/proto/RunRecord.java]: Yeah. Makes sense. Fixed the code to set to the empty map now instead of null.

**Comment by sagarkapare** [comment]
@anew Addressed comments and answered the questions. Please take a look again. Thanks!

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 3 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: But why delete it? You are gonna overwrite it in the next operation.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/ProgramWorkflowRunnerFactory.java (hunk 2 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/ProgramWorkflowRunnerFactory.java]: I see. I was not aware of these details.

**Comment by anew** [comment]
I still don't understand why the run record must be deleted before it can be overwritten. Other than that, LGTM.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: Ohh I see what you are saying. Yes, we are not changing the key also, so we do not need `deleteAll`. Removed in the last commit.

**Comment by sagarkapare** [comment]
Removed the redundant `deleteAll`. Build is running here http://builds.cask.co/browse/CDAP-RBT222-4. Will merge this once the build passes.

**Comment by sagarkapare** [comment]
Un-related test case failed which is failing on release branch as well. Merging this PR.

**Pull request merged**